sophie-kaleba/assignment2/MyBinaryTreeTest.java

```diff
@@ -0,0 +1,56 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+import java.util.Arrays;
+
+import org.junit.Before;
+import org.junit.Test;
+
+public class MyBinaryTreeTest {
+
+	public MyBinaryTree tree;
+
+	@Before
+	public final void init() {
+		this.tree = new MyBinaryTree(7);
+		MyBinaryTree.addNode(this.tree, new MyBinaryTree(3));
+		MyBinaryTree.addNode(this.tree, new MyBinaryTree(4));
+		MyBinaryTree.addNode(this.tree, new MyBinaryTree(2));
+	}
+
+	@Test
+	public final void addNewNodeTest() {
+		MyBinaryTree.addNode(this.tree, new MyBinaryTree(5));
+		int value = this.tree.getLeftNode().getRightNode().getNodeValue();
+		assertEquals(value,5);		
+	}
+
+	@Test
+	public final void printAncestorOKTest() {
+		MyBinaryTree.findAncestors(this.tree, 2, true, tree.getPath());
+		assertEquals(tree.getPath(), Arrays.asList(2,3,7));
+	}
+
+	@Test
+	public final void printAncestorRoot() {
+		MyBinaryTree.findAncestors(this.tree, 7, true, tree.getPath());
+		assertEquals(tree.getPath(), Arrays.asList(7));
+	}
+
+
+	@Test 
+	public final void printAncestorNodeNotFoundTest() {
+		assertFalse(MyBinaryTree.findAncestors(this.tree, 8, true, tree.getPath()));
+	}
+
```

Comment:
WHat happens if you have a duplicate node Could you add a test for that?

Reply:
if I have a duplicate node (in the LCA case, or when finding for an ancestor), the highest (ie, closest to the root) -"leftest" node will always be chosen.
I know that it is usually assumed that there are no duplicates (in this case, I should modify my "insert" method to make sure I am not adding a node already present in the tree). 
But if we assume there can be duplicates, what would be the best way to get around my issue? I was considering adding additional data for each node (a unique id maybe?) ?

Reply:
The simplest thing (which is often the best) is to document. Add a comment saying "in the event of duplicates the left most entry is taken" or something like that.

The other option is forbid your tree from taking duplicates, and enforce that somehow (which would need another data structure).